### DISCUSSION TITLE
Where is the code related to change controller in the firmware?

### DISCUSSION THREAD

**Discussion by cc299792458** [bitcraze/discussions#667] [Q&A] [answered]
Hi, 

I'd love to study the code of changing the controller dynamicly in the firmware. I found something in the stabilizer.c, L260-L264, it seems like the controller is modified by the variable 'controllerType'. But I failed to find the code to change the 'controllerType'. I wonder where is it?

I also know there is a api in crazyflie-lib-python to change the controller.

![Screenshot from 2023-04-19 13-42-13](https://user-images.githubusercontent.com/52767591/232977485-f9827c59-1ada-4eca-a333-2095fa178044.png)

Thank you very much!

**Comment by krichardsson** [comment, marked as answer]
Hi.

you can change controller with the [stabilizer.controller parameter](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#stabilizercontroller)

Implemented in the [firmware here](https://github.com/bitcraze/crazyflie-firmware/blob/9aaf568b5b137fb68e484f92390de87b77d08ab5/src/modules/src/stabilizer.c#L370)

[Example of setting the controller from the python lib can be found in autonomous_sequence_high_level.py for instance.
](https://github.com/bitcraze/crazyflie-lib-python/blob/6370208d61cd959c1009971558b53adb43e3a24b/examples/autonomy/autonomous_sequence_high_level.py#L114-L115)

**Reply by cc299792458** [reply]
Thank you! So the controller is changed by the parameter implementation. I got it!
